File: dsa-05-main/cpp/day16/BfsDfsAlgorithms.cpp
#include <iostream>
#include <stack>
#include <queue>
#include <cstring>
using namespace std;

class AdjMatrixNonWeightedGraph
{
private:
	int **mat;
	int vertCount, edgeCount;

public:
	AdjMatrixNonWeightedGraph(int vCount)
	{
		vertCount = vCount;
		edgeCount = 0;
		mat = new int *[vertCount];
		for (int i = 0; i < vertCount; i++)
		{
			mat[i] = new int[vertCount];
			for (int j = 0; j < vertCount; j++)
				mat[i][j] = 0;
		}
	}
	~AdjMatrixNonWeightedGraph()
	{
		for (int i = 0; i < vertCount; i++)
			delete[] mat[i];
		delete[] mat;
		mat = NULL;
		vertCount = 0;
		edgeCount = 0;
	}

	void accept()
	{
		cout << "Enter number of edges: ";
		cin >> edgeCount;
		cout << "Enter " << edgeCount << " edges -- src dest weight:" << endl;
		for (int i = 0; i < edgeCount; i++)
		{
			int src, dest;
			cin >> src >> dest;
			mat[src][dest] = 1;
			mat[dest][src] = 1; // for undirected graph src -> dest => dest -> src
		}
	}

	void display()
	{
		for (int i = 0; i < vertCount; i++)
		{
			cout << "v" << i;
			for (int j = 0; j < vertCount; j++)
				cout << "\t" << mat[i][j];
			cout << endl;
		}
	}

	void dfsTrav(int start)
	{
		bool *marked = new bool[vertCount];
		memset(marked, 0, vertCount * sizeof(bool));
		stack<int> s;
		cout << "DFS: ";
		// push start vertex on stack and mark it
		s.push(start);
		marked[start] = true;
		// while stack is not empty
		while (!s.empty())
		{
			// pop vertex from stack and visit it
			int trav = s.top();
			s.pop();
			cout << "\t" << trav << ", ";
			// put all its non-marked adjacent vertices on the stack & mark them
			for (int j = 0; j < vertCount; j++)
			{
				if (mat[trav][j] != 0 && marked[j] == false)
				{
					s.push(j);
					marked[j] = true;
				}
			}
		}
		cout << endl;
		delete[] marked;
	}

	bool dfsIsConnected()
	{
		bool *marked = new bool[vertCount];
		memset(marked, 0, vertCount * sizeof(bool));
		stack<int> s;
		int count = 0;

		int start = 0;
		s.push(start);
		marked[start] = true;
		count++;

		while (!s.empty())
		{
			int trav = s.top();
			s.pop();

			for (int j = 0; j < vertCount; j++)
			{
				if (mat[trav][j] == 1 && marked[j] == false)
				{
					s.push(j);
					marked[j] = true;
					count++;
					if (count == vertCount)
					{
						delete[] marked;
						return true;
					}
				}
			}
		}
		delete[] marked;
		return false;
	}

	void dfsSpanningTree(int start)
	{
		bool *marked = new bool[vertCount];
		memset(marked, 0, vertCount * sizeof(bool));
		stack<int> s;
		cout << "DFS Spanning Tree: " << endl;

		s.push(start);
		marked[start] = true;

		while (!s.empty())
		{
			int trav = s.top();
			s.pop();

			for (int j = 0; j < vertCount; j++)
			{
				if (mat[trav][j] == 1 && marked[j] == false)
				{
					s.push(j);
					marked[j] = true;
					cout << trav << " -> " << j << endl;
				}
			}
		}
		cout << endl;
		delete[] marked;
	}

	void bfsTrav(int start)
	{
		bool *marked = new bool[vertCount];
		memset(marked, 0, vertCount * sizeof(bool));
		queue<int> s;
		cout << "BFS: ";
		// push start vertex on queue and mark it
		s.push(start);
		marked[start] = true;
		// while queue is not empty
		while (!s.empty())
		{
			// pop vertex from queue and visit it
			int trav = s.front();
			s.pop();
			cout << "\t" << trav << ", ";
			// put all its non-marked adjacent vertices on the queue & mark them
			for (int j = 0; j < vertCount; j++)
			{
				if (mat[trav][j] != 0 && marked[j] == false)
				{
					s.push(j);
					marked[j] = true;
				}
			}
		}
		cout << endl;
		delete[] marked;
	}

	void bfsSpanningTree(int start)
	{
		bool *marked = new bool[vertCount];
		memset(marked, 0, vertCount * sizeof(bool));
		queue<int> s;
		cout << "BFS Spanning Tree: " << endl;

		s.push(start);
		marked[start] = true;

		while (!s.empty())
		{
			int trav = s.front();
			s.pop();

			for (int j = 0; j < vertCount; j++)
			{
				if (mat[trav][j] == 1 && marked[j] == false)
				{
					s.push(j);
					marked[j] = true;
					cout << trav << " -> " << j << endl;
				}
			}
		}
		cout << endl;
		delete[] marked;
	}

	void bfsPathLength(int start)
	{
		int *dist = new int[vertCount];
		bool *marked = new bool[vertCount];
		memset(marked, 0, vertCount * sizeof(bool));
		queue<int> s;
		s.push(start);
		marked[start] = true;
		dist[start] = 0;
		while (!s.empty())
		{
			int trav = s.front();
			s.pop();
			for (int j = 0; j < vertCount; j++)
			{
				if (mat[trav][j] == 1 && marked[j] == false)
				{
					s.push(j);
					marked[j] = true;
					dist[j] = dist[trav] + 1;
				}
			}
		}
		cout << "Single Source Path Length: " << endl;
		for (int i = 0; i < vertCount; i++)
			cout << "Distance of " << i << " from " << start << " = " << dist[i] << endl;
		cout << endl;
		delete[] dist;
		delete[] marked;
	}

	bool bfsIsBipartite()
	{
		const int NOCOLOR = 0, YELLOW = 1, GREEN = -1;
		int start = 0;
		int *color = new int[vertCount];
		memset(color, 0, vertCount * sizeof(int));
		for (int i = 0; i < vertCount; i++)
			color[i] = NOCOLOR;
		queue<int> s;
		s.push(start);
		color[start] = YELLOW;

		while (!s.empty())
		{
			int trav = s.front();
			s.pop();

			for (int j = 0; j < vertCount; j++)
			{
				if (mat[trav][j] == 1)
				{
					if (color[j] == NOCOLOR)
					{
						s.push(j);
						color[j] = -1 * color[trav];
					}
					else
					{
						if (color[j] == color[trav])
						{
							delete[] color;
							return false; // not bipartite graph
						}
					}
				}
			}
		}
		delete[] color;
		return true;
	}
};

int main()
{
	AdjMatrixNonWeightedGraph g(6);
	g.accept();
	g.display();
	g.dfsTrav(0);
	g.bfsTrav(0);
	if (g.dfsIsConnected())
		cout << "Graph is connected." << endl;
	else
		cout << "Graph is not connected." << endl;
	g.dfsSpanningTree(0);
	g.bfsSpanningTree(0);
	g.bfsPathLength(0);
	if (g.bfsIsBipartite())
		cout << "Graph is bi-partite." << endl;
	else
		cout << "Graph is not bi-partite." << endl;
	return 0;
}

/*
7
0 1
0 2
0 3
1 2
1 4 
3 4
3 5 
*/

File: dsa-05-main/python/day16/BfsDfsAlgorithms.py

class AdjMatrixNonWeightedGraph:


	def __init__(self, vCount):
		self.vertCount = vCount
		self.edgeCount = 0
		self.mat = []
		for i in range(self.vertCount):
			self.mat.append([0] * self.vertCount)
	

	def accept(self):
		self.edgeCount = int(input("Enter number of edges: "))
		print(f"Enter {self.edgeCount} edges -- src dest:")
		for i in range(self.edgeCount):
			(src_str, dest_str) = input().split()
			(src,dest) = (int(src_str), int(dest_str))
			self.mat[src][dest] = 1
			self.mat[dest][src] = 1 # for undirected graph src -> dest => dest -> src 


	def display(self):
		for i in range(self.vertCount):
			print(f"v{i}", end='')
			for j in range(self.vertCount):
				print("\t{}".format(self.mat[i][j]), end='')
			print()


	def dfsTrav(self, start):
		marked = [False] * self.vertCount
		s = list()
		print("DFS: ", end='')
		# push start vertex on stack and mark it
		s.append(start)
		marked[start] = True
		# while stack is not empty
		while s:
			# pop vertex from stack and visit it
			trav = s.pop()
			print(f"{trav}, ", end='')
			# put all its non-marked adjacent vertices on the stack & mark them
			for j in range(self.vertCount):
				if self.mat[trav][j]!=0 and marked[j]==False:
					s.append(j)
					marked[j] = True
		print()

	
	def dfsIsConnected(self):
		start = 0
		marked = [False] * self.vertCount
		count = 0
		s = list()
		s.append(start)
		marked[start] = True
		count = count + 1
		while s:
			trav = s.pop()
			for j in range(self.vertCount):
				if self.mat[trav][j]!=0 and marked[j]==False:
					s.append(j)
					marked[j] = True
					count = count + 1
					if count == self.vertCount:
						return True
		return False


	def dfsSpanningTree(self, start):
		marked = [False] * self.vertCount
		s = list()
		print("DFS Spanning Tree:")
		s.append(start)
		marked[start] = True
		while s:
			trav = s.pop()
			for j in range(self.vertCount):
				if self.mat[trav][j]!=0 and marked[j]==False:
					s.append(j)
					marked[j] = True
					print(f"{trav} -> {j}")
		print()


	def bfsTrav(self, start):
		marked = [False] * self.vertCount
		s = list()
		print("BFS: ", end='')
		# push start vertex on queue and mark it
		s.append(start)
		marked[start] = True
		# while queue is not empty
		while s:
			# pop vertex from queue and visit it
			trav = s.pop(0)
			print(f"{trav}, ", end='')
			# put all its non-marked adjacent vertices on the queue & mark them
			for j in range(self.vertCount):
				if self.mat[trav][j]!=0 and marked[j]==False:
					s.append(j)
					marked[j] = True
		print()


	def bfsSpanningTree(self, start):
		marked = [False] * self.vertCount
		s = list()
		print("BFS Spanning Tree:")
		s.append(start)
		marked[start] = True
		while s:
			trav = s.pop(0)
			for j in range(self.vertCount):
				if self.mat[trav][j]!=0 and marked[j]==False:
					s.append(j)
					marked[j] = True
					print(f"{trav} -> {j}")
		print()


	def bfsPathLength(self, start):
		dist = [0] * self.vertCount
		marked = [False] * self.vertCount
		s = list()
		s.append(start)
		marked[start] = True
		dist[start] = 0
		while s:
			trav = s.pop(0)
			for j in range(self.vertCount):
				if self.mat[trav][j]!=0 and marked[j]==False:
					s.append(j)
					marked[j] = True
					dist[j] = dist[trav] + 1
		print("Single Source Path Length: ")
		for i in range(self.vertCount):
			print(f"Distance of v{i} from v{start} = {dist[i]}")
		print()


	def bfsIsBipartite(self):
		NOCOLOR = 0
		YELLOW = 1
		GREEN = -1
		start = 0
		color = [NOCOLOR] * self.vertCount
		s = list()
		s.append(start)
		color[start] = YELLOW
		while s:
			trav = s.pop(0)
			for j in range(self.vertCount):
				if self.mat[trav][j]!=0:
					if color[j]==NOCOLOR:
						s.append(j)
						color[j] = -1 * color[trav]
					elif color[j]==color[trav]:
						return False
		return True


if __name__=="__main__":
	g = AdjMatrixNonWeightedGraph(6)
	g.accept()
	g.display()
	g.dfsTrav(0)
	g.bfsTrav(0)
	if g.dfsIsConnected():
		print("graph is connected.")
	else:
		print("graph is not connected.")
	g.dfsSpanningTree(0)
	g.bfsSpanningTree(0)
	if g.bfsIsBipartite():
		print("graph is bipartite.")
	else:
		print("graph is not bipartite.")
	g.bfsPathLength(0)


"""
7
0 1
0 2
0 3
1 2
1 4
3 4
3 5
"""
